llvm/docs/LangRef.rst

```diff
@@ -15694,6 +15698,47 @@ trapping or setting ``errno``.
 When specified with the fast-math-flag 'afn', the result may be approximated
 using a less accurate calculation.
 
+
+'``llvm.sincos.*``' Intrinsic
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Syntax:
+"""""""
+
+This is an overloaded intrinsic. You can use ``llvm.sincos`` on any
+floating-point or vector of floating-point type. Not all targets support
+all types however.
+
+::
+
+      declare { float, float }          @llvm.sincos.f32(float  %Val)
+      declare { double, double }        @llvm.sincos.f64(double %Val)
+      declare { x86_fp80, x86_fp80 }    @llvm.sincos.f80(x86_fp80  %Val)
+      declare { fp128, fp128 }          @llvm.sincos.f128(fp128 %Val)
+      declare { ppc_fp128, ppc_fp128 }  @llvm.sincos.ppcf128(ppc_fp128  %Val)
+      declare { <4 x float>, <4 x float> } @llvm.sincos.v4f32(<4 x float>  %Val)
+
+Overview:
+"""""""""
+
+The '``llvm.sincos.*``' intrinsics returns the sine and cosine of the operand.
+
+Arguments:
+""""""""""
+
+The argument is a :ref:`floating-point <t_floating>` or :ref:`vector <t_vector>`
+of floating-point values. Returns two values matching the argument type in a
+struct.
+
+Semantics:
+""""""""""
+
+This intrinsic is equivalent to a calling both :ref:`llvm.sin <t_llvm_sin>`
+and :ref:`llvm.cos <t_llvm_cos>` on the argument.
+
+The first result is the sine of the argument and the second result is the cosine
+of the argument.
```

Comment:
When specified with the fast-math-flag ‘afn’, the result may be approximated using a less accurate calculation.

sin and cos support fast-math-flags, but sincos does not?

Reply:
as a struct output, this is not an FPMathOperator

Reply:
But when I replace sincos by sin + cos then I get fast-math-flags, see IRTranslator changes.

Reply:
I think the definition of `FPMathOperator` would need to be changed for `sincos` (currently it does not allow struct values).

Reply:
`MachineInstr::copyFlagsFromInstruction(CI)` for the sincos intrinsic.

Reply:
You mean special-case `sincos` without updating the definition of `FPMathOperator`?

Reply:
I had a look into this, but it seems fast-math flags are fairly defendant on `FPMathOperator`.  It looks like a while ago it was extended to support ArrayTypes (https://reviews.llvm.org/D69161), so I've created a similar PR to extend it to support homogeneous literal struct types: https://github.com/llvm/llvm-project/pull/110506 (which would allow it for `llvm.sincos`).

Reply:
Added the note about `afn` now that #110506 has landed (though the actual approximation is not implemented in this patch).